pkg/persistence/redis/store.go

```diff
@@ -0,0 +1,166 @@
+package redis
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/goharbor/harbor-scanner-clair/pkg/etc"
+	"github.com/goharbor/harbor-scanner-clair/pkg/harbor"
+	"github.com/goharbor/harbor-scanner-clair/pkg/job"
+	"github.com/goharbor/harbor-scanner-clair/pkg/persistence"
+	"github.com/gomodule/redigo/redis"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/xerrors"
+	"time"
+)
+
+type redisStore struct {
+	cfg  etc.StoreConfig
+	pool redis.Pool
+}
+
+func NewStore(cfg etc.StoreConfig) persistence.Store {
+	return &redisStore{
+		cfg: cfg,
+		pool: redis.Pool{
+			Dial: func() (redis.Conn, error) {
+				return redis.DialURL(cfg.RedisURL)
+			},
+			MaxIdle:   cfg.PoolMaxIdle,
+			MaxActive: cfg.PoolMaxActive,
+			Wait:      true,
+		},
+	}
+}
+
+func (rs *redisStore) Save(scanJob job.ScanJob) error {
+	conn := rs.pool.Get()
+	defer rs.close(conn)
+
+	bytes, err := json.Marshal(scanJob)
+	if err != nil {
+		return xerrors.Errorf("marshalling scan job: %w", err)
+	}
+
+	key := rs.getKeyForScanJob(scanJob.ID)
+
+	log.WithFields(log.Fields{
+		"scan_job_id":     scanJob.ID,
+		"scan_job_status": scanJob.Status.String(),
+		"redis_key":       key,
+	}).Debug("Saving scan job")
+
+	_, err = conn.Do("SET", key, string(bytes))
+	if err != nil {
+		return xerrors.Errorf("saving scan job: %w", err)
+	}
+
+	return nil
+}
+
+func (rs *redisStore) saveWithExpire(scanJob job.ScanJob, expire time.Duration) error {
+	conn := rs.pool.Get()
+	defer rs.close(conn)
+
+	bytes, err := json.Marshal(scanJob)
+	if err != nil {
+		return xerrors.Errorf("marshalling scan job: %w", err)
+	}
+
+	key := rs.getKeyForScanJob(scanJob.ID)
+
+	log.WithFields(log.Fields{
+		"scan_job_id":     scanJob.ID,
+		"scan_job_status": scanJob.Status.String(),
+		"expire":          expire.String(),
+		"redis_key":       key,
+	}).Debug("Saving scan job with expire")
+
+	err = conn.Send("MULTI")
+	if err != nil {
+		return err
+	}
+	err = conn.Send("SET", key, string(bytes))
```

Comment:
`SET` with option `EX` can do the same work, no transaction needed. Check [SET](https://redis.io/commands/set)

Reply:
Thanks! Refactored to use EX arg.